### PR TITLE
Dev offline check fixed

### LINKED ISSUE #123: Save `UserPreference`s persistently and locally using `Datastore`

### LINKED ISSUE #128: Make Sure Offline Mode is functional

### PR DESCRIPTION
**Pull request by L-roro**
closes #128
closes #123

This pull request makes sure that offline mode doesn't crash. It also stores UserPreferences persistently using sharedPreferences. Unfortunately, I have struggled a lot trying to find a clean way of implementing all this while keeping the authentication context injection the way it was done. I explained this properly in a todo in authenticationModule.
I would gladly receive some advice concerning that in particular.

### CODE REVIEW

File: app/src/main/java/com/github/leuludyha/ibdb/di/AuthenticationModule.kt [hunk 1]
```diff
@@ -6,26 +6,39 @@ import com.github.leuludyha.domain.model.user.MainUser
...
+    @Singleton
     @Provides
     fun provideAuthenticationContext(libraryRepository: LibraryRepository): AuthenticationContext {
         return AuthenticationContext(
             MainUser(
-                Firebase.auth.currentUser?.uid!!,
-                Firebase.auth.currentUser?.displayName ?: "username",
-                Firebase.auth.currentUser?.photoUrl.toString(),
-                Firebase.auth.currentUser?.phoneNumber,
+                //TODO: Unfortunately cannot call Firebase directly from here (would fail if there was
```
Comment (WollfieGitHub) [inline app/src/main/java/com/github/leuludyha/ibdb/di/AuthenticationModule.kt:26]: As we discuss, that is unfortunate but we haven't found a cleaner solution yet...
`Constants.USER_NOT_FOUND` is nice tho..

File: app/src/main/java/com/github/leuludyha/ibdb/presentation/components/books/recommendations/RecommendationList.kt [hunk 2]
```diff
@@ -34,7 +39,11 @@ fun RecommendationList(
             work = it,
             displaySubjects = false,
             onClick = { clickedWork ->
-                navController.navigate(Screen.BookDetails.passBookId(clickedWork.id))
+                if (checkNetworkAvailable(context)) {
```
Comment (WollfieGitHub) [inline app/src/main/java/com/github/leuludyha/ibdb/presentation/components/books/recommendations/RecommendationList.kt:42]: This is nice :0

File: app/src/main/java/com/github/leuludyha/ibdb/presentation/screen/auth/signup/SignUpScreenViewModel.kt [hunk 3]
```diff
@@ -21,26 +23,63 @@ class SignUpScreenViewModel @Inject constructor(
      * Saves a boolean which indicates that the walk-through has been completed once
      * by the user. The boolean is persisted through sessions (Application restarts)
      */
-    fun rememberWalkThroughIsCompleted() {
+    fun rememberWalkThroughIsCompleted(context: Context) {
+        val sp = context.getSharedPreferences(
+            Constant.SIGN_UP_SHARED_PREFERENCES_KEY,
+            Context.MODE_PRIVATE
```
Comment (WollfieGitHub) [inline app/src/main/java/com/github/leuludyha/ibdb/presentation/screen/auth/signup/SignUpScreenViewModel.kt:29]: What does `MODE_PRIVATE` do ?

**Review by WollfieGitHub — APPROVED** [review]
You managed to make a hack somewhat clean :0
I think we can keep it this way for now

File: app/src/main/java/com/github/leuludyha/ibdb/presentation/screen/auth/signup/SignUpScreenViewModel.kt (hunk 3 above)
Comment (L-roro) [reply app/src/main/java/com/github/leuludyha/ibdb/presentation/screen/auth/signup/SignUpScreenViewModel.kt:29]: It prevents other applications from accessing these stored values.

**Pull request merged**